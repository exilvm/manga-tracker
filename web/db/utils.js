const format = require('pg-format');
const {
  NUMERIC_VALUE_OUT_OF_RANGE,
  INVALID_TEXT_REPRESENTATION,
} = require('pg-error-constants');
const dblog = require('debug')('db');

function generateEqualsColumns(o, availableColumns) {
  const cols = new Set(availableColumns);
  const sqlValues = [];
  const columns = [];
  const args = [];

  Object.keys(o)
    .filter(k => o[k] !== undefined && cols.has(k))
    .forEach((k, idx) => {
      sqlValues.push(`%I=$${idx+1}`);
      columns.push(k);
      args.push(o[k]);
    });

  const sqlCols = format(sqlValues.join(','), ...columns);

  return { sqlCols, args };
}
module.exports.generateEqualsColumns = generateEqualsColumns;

function handleError(err, res) {
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dblog(err.message);
    res.status(400).json({ error: 'Invalid data type given' });
    return;
  }

  if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: 'Number value out of range' });
    return;
  }

  console.error(err);
  res.status(500).json({ error: 'Internal server error' });
}
module.exports.handleError = handleError;
