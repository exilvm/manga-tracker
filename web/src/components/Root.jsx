/* eslint-disable */
import React from 'react';
import {Divider, Link, makeStyles, Typography} from '@material-ui/core';
import PropTypes from 'prop-types';
import TopBar from './TopBar';


const useStyles = makeStyles((theme) => ({
  root: {
    width: '100%',
    overflow: 'auto',
    minWidth: '400px',
    minHeight: '100vh',
    position: 'relative',
  },
  container: {
    paddingTop: theme.spacing(10),
  },
  divider: {
    marginTop: theme.spacing(5),
    marginBottom: theme.spacing(2),
  },
  footer: {
    bottom: theme.spacing(2),
    position: 'absolute',
    width: '100%',
    marginTop: theme.spacing(10),
  },
  copyright: {
    marginLeft: theme.spacing(3),
    bottom: theme.spacing(2),
  },
}));

function Copyright(props) {
  return (
    <Typography {...props}>
      { 'Copyright Â© '}
      <Link color='inherit' href='https://github.com/s0hv'>
        s0hv
      </Link>
      { ' ' }
      {new Date().getFullYear()}
      {'.'}
    </Typography>
  );
}

export default function Layout({ Component, pageProps, props }) {
  const {
    statusCode,
    activeTheme,
    user,
    setTheme,
  } = props;
  const classes = useStyles();

  if (statusCode !== 200) {
    return <Component {...pageProps} />;
  }

  return (
    <div className={classes.root}>
      <TopBar user={user} setTheme={setTheme} activeTheme={activeTheme} />
      <Component {...pageProps} isAuthenticated={Boolean(user)} />
      <div className={classes.container}>
        <footer className={classes.footer}>
          <Divider className={classes.divider} variant='middle' />
          <Copyright className={classes.copyright} />
        </footer>
      </div>
    </div>
  );
}

Layout.propTypes = {
  Component: PropTypes.elementType,
  // eslint-disable-next-line react/forbid-prop-types
  pageProps: PropTypes.object,
  props: PropTypes.shape({
    statusCode: PropTypes.number,
    activeTheme: PropTypes.number,
    user: PropTypes.object,
    setTheme: PropTypes.func,
  }),
};
