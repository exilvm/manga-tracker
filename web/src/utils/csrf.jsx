import { createContext, useContext } from 'react';

const CSRFContext = createContext(undefined);
export const CSRFProvider = CSRFContext.Provider;

export const useCSRF = () => useContext(CSRFContext);

/**
 * Returns the useCSRF token in the props property with the key _csrf
 * @param req
 * @returns {{props: {_csrf: String}}}
 */
export const csrfProps = ({ req }) => ({
  props: { _csrf: req.csrfToken() },
});

export const csrfHeader = (csrf) => ({
  'X-CSRF-Token': csrf,
});
