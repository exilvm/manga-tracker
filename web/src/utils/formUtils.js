export const showErrorAlways = ({
  meta: { submitError, dirtySinceLastSubmit, error },
}) => !!(((submitError && !dirtySinceLastSubmit) || error));
