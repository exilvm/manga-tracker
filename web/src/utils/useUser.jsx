import { createContext, useContext, useMemo } from 'react';


const UserContext = createContext(undefined);
export const UserProvider = UserContext.Provider;

export const useUser = () => {
  const user = useContext(UserContext);
  return useMemo(() => ({
    user,
    isAuthenticated: !!user,
    isAdmin: user && user.admin,
  }), [user]);
};
