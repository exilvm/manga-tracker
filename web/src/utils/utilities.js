import { format, formatDistanceToNowStrict } from 'date-fns';
import enLocale from 'date-fns/locale/en-GB';
import throttle from 'lodash.throttle';


// eslint-disable-next-line import/prefer-default-export
export const followUnfollow = (mangaId, serviceId) => {
  const url = serviceId ? `/api/user/follows?manga_id=${mangaId}&service_id=${serviceId}` :
                           `/api/user/follows?manga_id=${mangaId}`;
  return throttle((event) => {
    const target = event.target;
    switch (target.textContent.toLowerCase()) {
      case 'follow':
        fetch(url, { credentials: 'include', method: 'put' })
            .then(res => {
              if (res.status === 200) {
                target.textContent = 'Unfollow';
              }
            });
        break;

      case 'unfollow':
        fetch(url, { credentials: 'include', method: 'delete' })
            .then(res => {
              if (res.status === 200) {
                target.textContent = 'Follow';
              }
            });
        break;

      default:
        target.textContent = 'Follow';
    }
  }, 200, { trailing: false });
};

// This seems to be faster than a custom recursive function  according to my measurements
export function jsonSerializable(value) {
  return JSON.parse(JSON.stringify(value));
}

export function defaultDateFormat(date, ifUndefined='Unknown') {
  if (!date || Number.isNaN(date.getTime())) return ifUndefined;

  format(date, 'MMM do, HH:mm', { locale: enLocale });
}

export function defaultDateDistanceToNow(date, ifUndefined='Unknown') {
  if (!date || Number.isNaN(date.getTime())) return ifUndefined;

  return formatDistanceToNowStrict(date, { addSuffix: true });
}
