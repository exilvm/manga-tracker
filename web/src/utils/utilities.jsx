import { format, formatDistanceToNowStrict } from 'date-fns';
import enLocale from 'date-fns/locale/en-GB';
import throttle from 'lodash.throttle';

export const followUnfollow = (mangaId, serviceId) => {
  const url = serviceId ? `/api/user/follows?manga_id=${mangaId}&service_id=${serviceId}` :
    `/api/user/follows?manga_id=${mangaId}`;
  return throttle((event) => {
    const target = event.target;
    switch (target.textContent.toLowerCase()) {
      case 'follow':
        fetch(url, { credentials: 'include', method: 'put' })
          .then(res => {
            if (res.status === 200) {
              target.textContent = 'Unfollow';
            }
          });
        break;

      case 'unfollow':
        fetch(url, { credentials: 'include', method: 'delete' })
          .then(res => {
            if (res.status === 200) {
              target.textContent = 'Follow';
            }
          });
        break;

      default:
        target.textContent = 'Follow';
    }
  }, 200, { trailing: false });
};

// This seems to be faster than a custom recursive function  according to my measurements
export const jsonSerializable = (value) => JSON.parse(JSON.stringify(value));

function dateIsInvalid(date) {
  return !date || Number.isNaN(date.getTime()) || date.getTime() === 0;
}

export const defaultDateFormat = (date, ifUndefined='Unknown') => {
  if (dateIsInvalid(date)) return ifUndefined;

  return format(date, 'MMM do yyyy, HH:mm', { locale: enLocale });
};

export const defaultDateDistanceToNow = (date, ifUndefined='Unknown') => {
  if (dateIsInvalid(date)) return ifUndefined;

  return formatDistanceToNowStrict(date, { addSuffix: true });
};

export const noop = () => {};
