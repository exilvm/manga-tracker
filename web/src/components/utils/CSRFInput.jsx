import { Field } from 'react-final-form';
import { useCSRF } from '../../utils/csrf';

export default function CSRFInput() {
  const csrf = useCSRF();
  return (
    <Field
      name='_csrf'
      component='input'
      type='hidden'
      defaultValue={csrf}
    />
  );
}
