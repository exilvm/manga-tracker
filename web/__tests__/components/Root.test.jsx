import React from 'react';
import renderer from 'react-test-renderer';
import { createMount, createShallow } from '@material-ui/core/test-utils';

import Root from '../../src/components/Root';
import { adminUser, mockUTCDates, withUser } from '../utils';

const DummyComponent = () => <div />;

describe('Root component should render correctly', () => {
  mockUTCDates();

  // Replace current year with a fixed year
  jest.spyOn(Date.prototype, 'getFullYear')
    .mockImplementation(jest.fn(() => 2020));

  it('Should render with empty input', () => {
    const tree = renderer
      .create(<Root><DummyComponent /></Root>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Should render with empty input and correct status code', () => {
    const tree = renderer
      .create(<Root statusCode={200}><DummyComponent /></Root>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with data', () => {
    const tree = createMount()(
      <Root
        statusCode={200}
        activeTheme={1}
        setTheme={() => null}
      >
        <DummyComponent />
      </Root>
    );

    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with user', async () => {
    const elem = await withUser(
      adminUser,
      <Root
        statusCode={200}
        activeTheme={1}
        setTheme={() => null}
      >
        <DummyComponent />
      </Root>
    );
    const tree = createMount()(elem);

    expect(tree).toMatchSnapshot();
  });

  it('Should only return children on non 200 status code', () => {
    const wrapper = createShallow()(
      <Root
        statusCode={400}
        activeTheme={1}
        setTheme={() => null}
      >
        <DummyComponent />
      </Root>
    );
    // Check that dummy is the top level component
    expect(wrapper.find(DummyComponent).parent()).toHaveLength(0);
  });
});
