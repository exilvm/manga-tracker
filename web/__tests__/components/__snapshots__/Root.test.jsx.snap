// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root component should render correctly Should render with empty input 1`] = `<div />`;

exports[`Root component should render correctly Should render with empty input and correct status code 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="makeStyles-root-6"
  >
    <header
      className="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <h6
          className="MuiTypography-root makeStyles-title-8 MuiTypography-h6 MuiTypography-noWrap"
          onClick={[Function]}
        >
          Manga tracker
        </h6>
        <a
          href="/"
          onClick={[Function]}
          style={
            Object {
              "color": "inherit",
              "textDecoration": "none",
            }
          }
        >
          <button
            aria-controls="menu-appbar"
            aria-label="return to home page"
            className="MuiButtonBase-root MuiIconButton-root makeStyles-titleIcon-9 MuiIconButton-colorInherit"
            disabled={false}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
                />
              </svg>
            </span>
          </button>
        </a>
        <div
          className="makeStyles-grow-7"
        />
        <div
          className="makeStyles-search-15"
          onKeyDown={[Function]}
        >
          <div
            className="MuiInputBase-root makeStyles-inputRoot-17 MuiInputBase-adornedEnd"
            onClick={[Function]}
          >
            <input
              aria-label="search"
              autoCapitalize="none"
              autoComplete="off"
              className="MuiInputBase-input makeStyles-inputInput-18 MuiInputBase-inputAdornedEnd"
              id="title-search"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Search…"
              spellCheck={false}
              type="text"
            />
            <button
              className="MuiButtonBase-root MuiIconButton-root"
              disabled={false}
              onBlur={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-loginButton-11"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Login
          </span>
        </button>
        <button
          aria-label="Switch theme"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root makeStyles-menuItemIcon-13"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9 2c-1.05 0-2.05.16-3 .46 4.06 1.27 7 5.06 7 9.54 0 4.48-2.94 8.27-7 9.54.95.3 1.95.46 3 .46 5.52 0 10-4.48 10-10S14.52 2 9 2z"
              />
            </svg>
          </span>
        </button>
      </div>
    </header>
  </div>
  <div />
  <div
    className="makeStyles-container-2"
  >
    <footer
      className="makeStyles-footer-4"
    >
      <hr
        className="MuiDivider-root makeStyles-divider-3 MuiDivider-middle"
      />
      <p
        className="MuiTypography-root makeStyles-copyright-5 MuiTypography-body1"
      >
        Copyright © 
        <a
          className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
          href="https://github.com/s0hv"
          onBlur={[Function]}
          onFocus={[Function]}
        >
          s0hv
        </a>
         
        2020
        .
      </p>
    </footer>
  </div>
</div>
`;

exports[`Root component should render correctly should render correctly with data 1`] = `"<div class=\\"makeStyles-root-1\\"><div class=\\"makeStyles-root-6\\"><header class=\\"MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4\\"><div class=\\"MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters\\"><h6 class=\\"MuiTypography-root makeStyles-title-8 MuiTypography-h6 MuiTypography-noWrap\\">Manga tracker</h6><a style=\\"text-decoration: none;\\" href=\\"/\\"><button class=\\"MuiButtonBase-root MuiIconButton-root makeStyles-titleIcon-9 MuiIconButton-colorInherit\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"return to home page\\" aria-controls=\\"menu-appbar\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button></a><div class=\\"makeStyles-grow-7\\"></div><div class=\\"makeStyles-search-15\\"><div class=\\"MuiInputBase-root makeStyles-inputRoot-17 MuiInputBase-adornedEnd\\"><input autocomplete=\\"off\\" id=\\"title-search\\" placeholder=\\"Search…\\" type=\\"text\\" aria-label=\\"search\\" spellcheck=\\"false\\" autocapitalize=\\"none\\" class=\\"MuiInputBase-input makeStyles-inputInput-18 MuiInputBase-inputAdornedEnd\\" value=\\"\\"><button class=\\"MuiButtonBase-root MuiIconButton-root\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div><div class=\\"makeStyles-profileIcon-10\\"><button class=\\"MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"account of current user\\" aria-controls=\\"menu-appbar\\" aria-haspopup=\\"true\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root MuiSvgIcon-fontSizeLarge\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button></div></div></header></div><div></div><div class=\\"makeStyles-container-2\\"><footer class=\\"makeStyles-footer-4\\"><hr class=\\"MuiDivider-root makeStyles-divider-3 MuiDivider-middle\\"><p class=\\"MuiTypography-root makeStyles-copyright-5 MuiTypography-body1\\">Copyright © <a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit\\" href=\\"https://github.com/s0hv\\">s0hv</a> 2020.</p></footer></div></div>"`;
